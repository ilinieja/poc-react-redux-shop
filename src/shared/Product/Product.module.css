@import "vars.css";

.container {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    border: 1px solid var(--color-primary);
    padding: 20px;
    gap: 20px;
}

.row {
    display: flex;
    flex-direction: row;
    position: relative;
}

.title {
    font-size: 4rem;
    text-align: center;
    font-weight: inherit;
    margin: 0;
}

.subtitle {
    position: absolute;
    font-size: 1.4rem;
    bottom: 4px;
    left: calc(100% + 4px); 
}

.text {
    font-size: 1rem;
    text-align: center;
}
