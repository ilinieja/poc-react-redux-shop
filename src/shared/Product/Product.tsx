import { EntityId } from "@reduxjs/toolkit";
import classNames from "classnames";
import { useSelector } from "react-redux";
import { selectCartProductById, selectProductById } from "store/selectors";

import styles from "./Product.module.css";
import { ProductActions } from "./ProductActions/ProductActions";
import { ProductAdminActions } from "./ProductAdminActions/ProductAdminActions";
import { ProductRuleBadge } from "./ProductRuleBadge/ProductRuleBadge";

export interface ProductProps {
  productId: EntityId;
  className?: string;
  isAdmin?: boolean;
}

export function Product({ className, productId, isAdmin }: ProductProps) {
  const product = useSelector(selectProductById(productId));
  const cartProduct = useSelector(selectCartProductById(productId));

  if (!product) {
    // TODO: log warning from here when there's logger.
    console.warn(`No product for id: ${productId}`);
    return <></>;
  }

  return (
    <div className={classNames(styles.container, className)}>
      <ProductRuleBadge className={styles.badge} productId={productId} />
      <div className={styles.row}>
        <h2 className={styles.title}>{product.id}</h2>
        {!isAdmin && (
          <span className={styles.subtitle}>{cartProduct?.quantity || ""}</span>
        )}
      </div>
      <span className={styles.text}>{product.description ?? ""}</span>
      {isAdmin ? (
        <ProductAdminActions
          productId={productId}
          price={product.price}
        ></ProductAdminActions>
      ) : (
        <ProductActions
          productId={productId}
          price={product.price}
          quantity={cartProduct?.quantity ?? 0}
        ></ProductActions>
      )}
    </div>
  );
}
