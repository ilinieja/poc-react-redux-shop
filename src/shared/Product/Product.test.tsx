import { renderWithProviders, TEST_IDS } from "testing/testing";
import { Product } from "./Product";
import { fireEvent, screen } from "@testing-library/react";

describe("default mode", () => {
  it("renders not-bought product", () => {
    expect(
      renderWithProviders(<Product productId="X" />).baseElement
    ).toMatchSnapshot();
  });

  it("renders bought product", () => {
    expect(
      renderWithProviders(<Product productId="C" />).baseElement
    ).toMatchSnapshot();
  });

  it("renders bought product with special offer", () => {
    expect(
      renderWithProviders(<Product productId="A" />).baseElement
    ).toMatchSnapshot();
  });

  it("renders not-bought product with special offer", () => {
    expect(
      renderWithProviders(<Product productId="B" />).baseElement
    ).toMatchSnapshot();
  });

  it("updates product quantity when it's decremented", async () => {
    renderWithProviders(<Product productId="A" />);

    expect(
      screen.getByTestId(TEST_IDS.productLetterBoughtQuantity).textContent
    ).toMatchSnapshot();

    fireEvent.click(screen.getByTestId(TEST_IDS.productDecrementButton));

    expect(
      screen.getByTestId(TEST_IDS.productLetterBoughtQuantity).textContent
    ).toMatchSnapshot();
  });

  it("updates product quantity when it's incremented", async () => {
    renderWithProviders(<Product productId="A" />);

    expect(
      screen.getByTestId(TEST_IDS.productLetterBoughtQuantity).textContent
    ).toMatchSnapshot();

    fireEvent.click(screen.getByTestId(TEST_IDS.productIncrementButton));

    expect(
      screen.getByTestId(TEST_IDS.productLetterBoughtQuantity).textContent
    ).toMatchSnapshot();
  });

  it("updates button when product is bought", async () => {
    const rendered = renderWithProviders(<Product productId="Y" />);

    expect(screen.queryByTestId(TEST_IDS.productBuyButton)).toBeTruthy();
    expect(screen.queryByTestId(TEST_IDS.productIncrementButton)).toBeFalsy();
    expect(screen.queryByTestId(TEST_IDS.productDecrementButton)).toBeFalsy();

    fireEvent.click(screen.getByTestId(TEST_IDS.productBuyButton));

    expect(screen.queryByTestId(TEST_IDS.productBuyButton)).toBeFalsy();
    expect(screen.queryByTestId(TEST_IDS.productIncrementButton)).toBeTruthy();
    expect(screen.queryByTestId(TEST_IDS.productDecrementButton)).toBeTruthy();
  });
});

describe("admin mode", () => {
  it("renders not-bought product", () => {
    expect(
      renderWithProviders(<Product productId="X" isAdmin={true} />).baseElement
    ).toMatchSnapshot();
  });

  it("renders bought product", () => {
    expect(
      renderWithProviders(<Product productId="C" isAdmin={true} />).baseElement
    ).toMatchSnapshot();
  });

  it("renders bought product with special offer", () => {
    expect(
      renderWithProviders(<Product productId="A" isAdmin={true} />).baseElement
    ).toMatchSnapshot();
  });

  it("renders not-bought product with special offer", () => {
    expect(
      renderWithProviders(<Product productId="B" isAdmin={true} />).baseElement
    ).toMatchSnapshot();
  });
});

it("doesn't render non-existing product", () => {
  expect(
    renderWithProviders(<Product productId="?" />).baseElement
  ).toMatchSnapshot();
});
