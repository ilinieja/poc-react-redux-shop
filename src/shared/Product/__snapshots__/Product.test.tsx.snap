// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin mode renders bought product 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <h2
          class="title"
          data-testid="productLetterTitle"
        >
          C
        </h2>
      </div>
      <span
        class="text"
      >
        Awesome letter, I use it a lot and recommend
      </span>
      <div
        class="container"
      >
        <h3
          class="title"
        >
          20
        </h3>
        <div
          class="row"
        >
          <button
            class="button button"
          >
            Delete
          </button>
          <button
            class="button button widePadding"
          >
            Edit
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`admin mode renders bought product with special offer 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        class="container badge"
      >
        <div
          class="ribbon"
        >
          <h4
            class="text"
          >
            3
             for 
            130
          </h4>
        </div>
      </div>
      <div
        class="row"
      >
        <h2
          class="title"
          data-testid="productLetterTitle"
        >
          A
        </h2>
      </div>
      <span
        class="text"
      >
        Awesome letter, I use it a lot and recommend
      </span>
      <div
        class="container"
      >
        <h3
          class="title"
        >
          50
        </h3>
        <div
          class="row"
        >
          <button
            class="button button"
          >
            Delete
          </button>
          <button
            class="button button widePadding"
          >
            Edit
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`admin mode renders not-bought product 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <h2
          class="title"
          data-testid="productLetterTitle"
        >
          X
        </h2>
      </div>
      <span
        class="text"
      >
        Awesome letter, I use it a lot and recommend
      </span>
      <div
        class="container"
      >
        <h3
          class="title"
        >
          15
        </h3>
        <div
          class="row"
        >
          <button
            class="button button"
          >
            Delete
          </button>
          <button
            class="button button widePadding"
          >
            Edit
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`admin mode renders not-bought product with special offer 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        class="container badge"
      >
        <div
          class="ribbon"
        >
          <h4
            class="text"
          >
            2
             for 
            45
          </h4>
        </div>
      </div>
      <div
        class="row"
      >
        <h2
          class="title"
          data-testid="productLetterTitle"
        >
          B
        </h2>
      </div>
      <span
        class="text"
      >
        Awesome letter, I use it a lot and recommend
      </span>
      <div
        class="container"
      >
        <h3
          class="title"
        >
          30
        </h3>
        <div
          class="row"
        >
          <button
            class="button button"
          >
            Delete
          </button>
          <button
            class="button button widePadding"
          >
            Edit
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`default mode renders bought product 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <h2
          class="title"
          data-testid="productLetterTitle"
        >
          C
        </h2>
        <span
          class="subtitle"
          data-testid="productLetterBoughtQuantity"
        >
          5
        </span>
      </div>
      <span
        class="text"
      >
        Awesome letter, I use it a lot and recommend
      </span>
      <div
        class="container"
      >
        <h3
          class="title"
        >
          20
        </h3>
        <div
          class="buttonGroup"
        >
          <button
            class="button button"
            data-testid="productDecrementButton"
          >
            -
          </button>
          <button
            class="button button bold"
            data-testid="productIncrementButton"
          >
            +
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`default mode renders bought product with special offer 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        class="container badge"
      >
        <div
          class="ribbon"
        >
          <h4
            class="text"
          >
            3
             for 
            130
          </h4>
        </div>
      </div>
      <div
        class="row"
      >
        <h2
          class="title"
          data-testid="productLetterTitle"
        >
          A
        </h2>
        <span
          class="subtitle"
          data-testid="productLetterBoughtQuantity"
        >
          3
        </span>
      </div>
      <span
        class="text"
      >
        Awesome letter, I use it a lot and recommend
      </span>
      <div
        class="container"
      >
        <h3
          class="title"
        >
          50
        </h3>
        <div
          class="buttonGroup"
        >
          <button
            class="button button"
            data-testid="productDecrementButton"
          >
            -
          </button>
          <button
            class="button button bold"
            data-testid="productIncrementButton"
          >
            +
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`default mode renders not-bought product 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <h2
          class="title"
          data-testid="productLetterTitle"
        >
          X
        </h2>
        <span
          class="subtitle"
          data-testid="productLetterBoughtQuantity"
        />
      </div>
      <span
        class="text"
      >
        Awesome letter, I use it a lot and recommend
      </span>
      <div
        class="container"
      >
        <h3
          class="title"
        >
          15
        </h3>
        <button
          class="button button"
          data-testid="productBuyButton"
        >
          Buy
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`default mode renders not-bought product with special offer 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        class="container badge"
      >
        <div
          class="ribbon"
        >
          <h4
            class="text"
          >
            2
             for 
            45
          </h4>
        </div>
      </div>
      <div
        class="row"
      >
        <h2
          class="title"
          data-testid="productLetterTitle"
        >
          B
        </h2>
        <span
          class="subtitle"
          data-testid="productLetterBoughtQuantity"
        />
      </div>
      <span
        class="text"
      >
        Awesome letter, I use it a lot and recommend
      </span>
      <div
        class="container"
      >
        <h3
          class="title"
        >
          30
        </h3>
        <button
          class="button button"
          data-testid="productBuyButton"
        >
          Buy
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`default mode updates product quantity when it's decremented 1`] = `"3"`;

exports[`default mode updates product quantity when it's decremented 2`] = `"2"`;

exports[`default mode updates product quantity when it's incremented 1`] = `"3"`;

exports[`default mode updates product quantity when it's incremented 2`] = `"4"`;

exports[`doesn't render non-existing product 1`] = `
<body>
  <div />
</body>
`;
